File: 13. Roman to Integer.py
class Solution:
    def romanToInt(self, s: str) -> int:
        '''
        make a dict for look up, and then check some conditions for subtracting some number
        '''
        toret = 0
        dict = {
            'I' : 1,
            'V' : 5,
            'X' : 10,
            'L' : 50,
            'C' : 100,
            'D' : 500,
            'M' : 1000
        }
        for i, char in enumerate(s):
            toret += dict[char]
            if i != len(s) - 1:
                if char == 'C' and (s[i+1] == 'D' or s[i+1] == 'M'):
                    toret -= 200
                elif char == 'X' and (s[i+1] == 'L' or s[i+1] == 'C'):
                    toret -= 20
                elif char == 'I' and (s[i+1] == 'V' or s[i+1] == 'X'):
                    toret -= 2
        return toret

File: 13. Roman to Integer.java
class Solution {
    public int romanToInt(String s) {
        int res = 0;
        for (int i=s.length()-1; i>=0; i--){
            char c = s.charAt(i);
            switch (c) 
            {
                case 'I':
                    res += res < 5 ? 1 : -1;
                    break;
                case 'V':
                    res += res < 10 ? 5 : -5;
                    break;
                case 'X':
                    res += res < 50 ? 10: -10;
                    break;
                case 'L':
                    res += res < 100 ? 50: -50;
                    break;
                case 'C':
                    res += res < 500 ? 100: -100;
                    break;
                case 'D':
                    res += res < 1000? 500 : -500;
                    break;
                case 'M':
                    res += res < 5000? 1000 : -1000;
                    break;
            }
        }
        return res;
    }
}